### DISCUSSION TITLE
[Feature Request] Some Popular Emulators preconfigured as "Launchers"?

### DISCUSSION THREAD

**Discussion by Darthagnon** [MikeMaximus/gbm#224] [Ideas]
Noticed some of the latest work going into launcher support. I was wondering if it would be possible for a few popular emulators to be included as "Launchers"?

Here is some technical info:

Platform: | Emulator: | Save file location: |
--------- | -------- | -------------- |
Sony PSP | [PPSSPP](https://ppsspp.org/) | ```%UserProfile%\Documents\PPSSPP\PSP\SAVEDATA\[Game ID]``` (savestates in ```\PSP\PPSSPP_STATE\[GameID].[numbers].ppst```) |
Nintendo 3DS | [Citra](https://citra-emu.org/) | Savedata: ```%AppData%\Citra\sdmc\Nintendo 3DS\00000000000000000000000000000000\00000000000000000000000000000000\title\00040000\[Game ID]```, Extra Data (for some games, this is also savedata): ```%AppData%\Citra\sdmc\Nintendo 3DS\00000000000000000000000000000000\00000000000000000000000000000000\extdata\00000000\[Game ID]``` |
Nintendo Wii/Gamecube | [Dolphin](https://dolphin-emu.org/) | ```%UserProfile%\Documents\Dolphin Emulator\Wii\title\[Game ID]``` |
Sony PlayStation 3 | [RPCS3](https://rpcs3.net/) | ```[rpcs3.exe dir]\dev_hdd0\home\00000001\savedata\[Game ID]``` |

Other notable emulators that I haven't tested yet, but will update the list as I am able:

Platform | Emulator: | Save file location: |
--------- | ---------- | -------------------- | 
All Platforms | [RetroArch](https://www.retroarch.com/) |
Nintendo Switch | [Yuzu](https://yuzu-emu.org/) |
Nintendo Switch | [Ryujinx](https://ryujinx.org/) |
MS-DOS | [DOSBox](https://www.dosbox.com/) |
Sony PlayStation 2 | [PCSX2](https://pcsx2.net/) |
Sony PlayStation/PSOne/PSX | [Duckstation](https://github.com/stenzek/duckstation) |
Nintendo GameBoy + GameBoy Advance | [visualboyadvance-m](https://github.com/visualboyadvance-m/visualboyadvance-m) |
Nintendo GameBoy Advance | [NanoBoyAdvance](https://github.com/fleroviux/NanoboyAdvance/) |
Nintendo WiiU | [Cemu](http://cemu.info/) |
Nintendo 64 | [Project64](https://www.pj64-emu.com/) |
Nintendo 64 | [mupen64plus](http://mupen64plus.org/) |

**Comment by Darthagnon** [comment]
Oh, wow, the formatting looks bad on that first table. Not sure how to fix it, though. I used standard Markdown...

**Reply by Darthagnon** [reply]
Note to self: add table column that indicates whether the emulator is portable, and/or how to enable portable mode (e.g. Citra)

**Comment by MikeMaximus** [comment]
That's a good idea to consider, since v127 now supports external launchers that use their own executable + parameters to launch a game.

The issue with pre-configuring executable based launchers is that GBM needs to know where they are, as opposed to URI based launchers.  For example, v127 will now preconfigure GOG Galaxy, but only if it's installed to the default location.  I could possibly leverage GBM's existing executable search or ask the user where it is when it's not installed the default location.

**Reply by Darthagnon** [reply]
Asking the user where it is installed is a good idea, when they set up an emulator as a launcher with GBM, as a lot of these emulators are portable software, or have portable options.

**Comment by MikeMaximus** [comment]
I just wanted to post an update on this feature.  There won't be any additions for emulator support in the v127.  

I've been mulling over some changes to how GBM handles launchers and external game ids that I'd like to take care of first, which will make this feature better when I get around to it.

**Reply by Darthagnon** [reply]
Sounds good! I haven't had time to research more emulator save paths yet, either. 

I should probably do an update for the GBM offline manual soon, too.

**Reply by MikeMaximus** [reply]
> I should probably do an update for the GBM offline manual soon, too.

It might be best to wait on that until I do the manual updates for the launcher changes in v127. 😄
